include/quicly/streambuf.h

```diff
@@ -29,31 +29,113 @@ extern "C" {
 #include <assert.h>
 #include <stddef.h>
 #include <stdint.h>
+#include <string.h>
 #include "picotls.h"
 #include "quicly.h"
 
+typedef struct st_quicly_sendbuf_vec_t quicly_sendbuf_vec_t;
+
+/**
+ * Callback that flattens the contents of an iovec.
+ * @param dst the destination
+ * @param off offset within the iovec from where serialization should happen
+ * @param len number of bytes to serialize
+ * @return 0 if successful, otherwise an error code
+ */
+typedef int (*quicly_sendbuf_flatten_vec_cb)(quicly_sendbuf_vec_t *vec, void *dst, size_t off, size_t len);
+/**
+ * An optional callback that is called when an iovec is discarded
+ */
+typedef void (*quicly_sendbuf_discard_vec_cb)(quicly_sendbuf_vec_t *vec);
+
+typedef struct st_quicly_streambuf_sendvec_callbacks_t {
+    quicly_sendbuf_flatten_vec_cb flatten_vec;
+    quicly_sendbuf_discard_vec_cb discard_vec;
+} quicly_streambuf_sendvec_callbacks_t;
+
+struct st_quicly_sendbuf_vec_t {
+    const quicly_streambuf_sendvec_callbacks_t *cb;
+    size_t len;
+    void *cbdata;
+};
+
+typedef struct st_quicly_sendbuf_t {
+    struct {
+        quicly_sendbuf_vec_t *entries;
+        size_t size, capacity;
+    } vecs;
+    size_t off_in_first_vec;
+    uint64_t bytes_written;
+} quicly_sendbuf_t;
+
+static void quicly_sendbuf_init(quicly_sendbuf_t *sb);
+void quicly_sendbuf_dispose(quicly_sendbuf_t *sb);
```

Comment:
Nit: Can you add comments to all of these function declarations?

Reply:
Done in 8ed67be.